src/System.Collections.Immutable/tests/ImmutableArrayTest.cs

```diff
@@ -479,630 +577,1045 @@ public void LastIndexOf()
+        public void RemoveRangeIndexLength(IEnumerable<int> source, int index, int length)
+        {
+            var array = source.ToImmutableArray();
+            var expected = source.Take(index).Concat(source.Skip(index + length));
+            Assert.Equal(expected, array.RemoveRange(index, length));
+        }
+
+        public static IEnumerable<object[]> RemoveRangeIndexLengthData()
+        {
+            yield return new object[] { s_empty, 0, 0 };
+            yield return new object[] { s_oneElement, 1, 0 };
+            yield return new object[] { s_oneElement, 0, 1 };
+            yield return new object[] { s_oneElement, 0, 0 };
+            yield return new object[] { new[] { 1, 2, 3, 4 }, 0, 2 };
+            yield return new object[] { new[] { 1, 2, 3, 4 }, 1, 2 };
+            yield return new object[] { new[] { 1, 2, 3, 4 }, 2, 2 };
+        }
+
+        [Theory]
+        [MemberData(nameof(Int32EnumerableData))]
+        public void RemoveRangeIndexLengthInvalid(IEnumerable<int> source)
         {
-            Assert.Equal(new[] { 5 }, s_oneElement.Replace(1, 5));
+            var array = source.ToImmutableArray();
 
-            Assert.Equal(new[] { 6, 2, 3 }, s_manyElements.Replace(1, 6));
-            Assert.Equal(new[] { 1, 6, 3 }, s_manyElements.Replace(2, 6));
-            Assert.Equal(new[] { 1, 2, 6 }, s_manyElements.Replace(3, 6));
+            Assert.Throws<ArgumentOutOfRangeException>("index", () => array.RemoveRange(-1, 1));
+            Assert.Throws<ArgumentOutOfRangeException>("index", () => array.RemoveRange(array.Length + 1, 1));
+            Assert.Throws<ArgumentOutOfRangeException>("length", () => array.RemoveRange(0, -1));
+            Assert.Throws<ArgumentOutOfRangeException>("length", () => array.RemoveRange(0, array.Length + 1));
+        }
 
-            Assert.Equal(new[] { 1, 2, 3, 4 }, ImmutableArray.Create(1, 3, 3, 4).Replace(3, 2));
+        [Theory]
+        [InlineData(-1, 0, Skip = "TODO. See comment in ImmutableArray<T>.RemoveRange(int, int).")]
+        [InlineData(0, -1)]
+        [InlineData(0, 0)]
+        [InlineData(1, -1)]
+        public void RemoveRangeIndexLengthDefaultInvalid(int index, int length)
+        {
+            Assert.Throws<NullReferenceException>(() => s_emptyDefault.RemoveRange(index, length));
         }
 
-        [Fact]
-        public void ReplaceWithEqualityComparerTest()
+        [Theory]
+        [MemberData(nameof(RemoveRangeEnumerableData))]
+        public void RemoveRangeEnumerable(IEnumerable<int> source, IEnumerable<int> items, IEqualityComparer<int> comparer)
         {
-            var updatedArray = s_manyElements.Replace(2, 10, null);
-            Assert.Equal(new[] { 1, 10, 3 }, updatedArray);
+            var array = source.ToImmutableArray();
+            IEnumerable<int> expected = items.Aggregate(
+                seed: source.ToImmutableArray(),
+                func: (a, i) => a.Remove(i, comparer));
```

Comment:
Self-note: This is the cleanest way to calculate `expected` but it seems uncomfortably close to the code I'm actually testing.